src/storage/mod.rs

```diff
@@ -568,62 +488,187 @@ impl Storage {
+                            thread_ctx.collect_key_reads(CMD, r.len() as u64);
+                            r
+                        });
+
+                    thread_ctx.collect_scan_count(CMD, &statistics);
+                    thread_ctx.collect_read_flow(ctx.get_region_id(), &statistics);
+
+                    result
+                })
+        }) {
+            Ok(val) => box val,
+            Err(_) => box future::err(Error::SchedTooBusy),
+        }
     }
 
+    /// Scan a range starting with `start_key` up to `limit` rows from the snapshot.
     pub fn async_scan(
         &self,
         ctx: Context,
         start_key: Key,
         limit: usize,
         start_ts: u64,
         options: Options,
-        callback: Callback<Vec<Result<KvPair>>>,
-    ) -> Result<()> {
-        let cmd = Command::Scan {
-            ctx: ctx,
-            start_key: start_key,
-            limit: limit,
-            start_ts: start_ts,
-            options: options,
-        };
-        let tag = cmd.tag();
-        self.schedule(cmd, StorageCb::KvPairs(callback))?;
-        KV_COMMAND_COUNTER_VEC.with_label_values(&[tag]).inc();
-        Ok(())
+    ) -> Box<Future<Item = Vec<Result<KvPair>>, Error = Error> + Send> {
+        static CMD: &'static str = "scan";
+        let engine = self.get_engine();
+        let priority = readpool::Priority::from(ctx.get_priority());
+
+        match self.read_pool.future_execute(priority, move |ctxd| {
+            let _timer = {
+                let ctxd = ctxd.clone();
+                let mut thread_ctx = ctxd.current_thread_context_mut();
+                thread_ctx.collect_command_count(CMD, priority, false);
+                thread_ctx.collect_command_duration(CMD)
+            };
+
+            engine
+                .future_snapshot(&ctx)
+                // map storage::engine::Error -> storage::txn::Error -> storage::Error
+                .map_err(txn::Error::from)
+                .map_err(Error::from)
+                .and_then(move |snapshot: Box<Snapshot>| {
+                    let mut thread_ctx = ctxd.current_thread_context_mut();
+                    let _t_process = thread_ctx.collect_processing_read_duration(CMD);
+
+                    let snap_store = SnapshotStore::new(
+                        snapshot,
+                        start_ts,
+                        ctx.get_isolation_level(),
+                        !ctx.get_not_fill_cache(),
+                    );
+                    snap_store
+                        .scanner(ScanMode::Forward, options.key_only, None, None)
+                        .and_then(|mut scanner| {
+                            let res = scanner.scan(start_key, limit);
+                            let statistics = scanner.get_statistics();
+                            thread_ctx.collect_scan_count(CMD, statistics);
+                            thread_ctx.collect_read_flow(ctx.get_region_id(), statistics);
```

Comment:
ditto, could we merge `collect_scan_count` && `collect_read_flow ` into a single function?

Reply:
If they always come together we can merge them. I will take a deeper look at this.

Reply:
Seems that we cannot merge this two. i.e. for async_raw_scan, it has scan_count but does not has read flow.

Reply:
Why `async_raw_scan` do not has read flow?